File: King of Modules/Backups/16-03-26 v2.0.1 --- Python MT5 e Backtest/Strategies/RSIExhaustion.py
import pandas as pd
 
from Config import cfg, Signal
from Indicators import (
    get_rsi,
    get_keltner,
    get_kama_color,
    get_kama_values,
    get_atr_value,
)
from MeanReversion import _update_mr_indicators   # shared helper
 
 
# Default multiplier (not in Inputs.mqh in the original source)
_ATR_RE_TREND_MULTIPLIER: float = 1.0
 
 
# ---------------------------------------------------------------------------
# ATR Compression Filter — PassREATRCompressionFilter
# ---------------------------------------------------------------------------
 
def _pass_re_atr_compression_filter(rates: pd.DataFrame) -> bool:
    """
    Allow entry when ATR(14) * multiplier < ATR(50).
    Mirrors PassREATRCompressionFilter().
    """
    atr_fast = get_atr_value(rates, cfg.atr_period)
    atr_slow = get_atr_value(rates, 50)
    if atr_fast is None or atr_slow is None:
        return False
 
    return atr_fast * _ATR_RE_TREND_MULTIPLIER < atr_slow
 
 
# ---------------------------------------------------------------------------
# Exhaustion Detection — DetectRSIExhaustion
# ---------------------------------------------------------------------------
 
def _detect_rsi_exhaustion(
    cur_rsi:    float,
    upper_k:    float,
    lower_k:    float,
    kama_color: int,
    close1:     float,
) -> Signal:
    """
    Mirrors DetectRSIExhaustion().
 
    Note: the logic uses OR between RSI and Keltner conditions (not AND
    as in Mean Reversion). This is preserved exactly as written in the source.
 
    Oversold  + KAMA bearish(1) → SELL  (fading exhausted move down)
    Overbought + KAMA bullish(2) → BUY  (fading exhausted move up)
    """
    if (cur_rsi < cfg.rsi_oversold or close1 < lower_k) and kama_color == 1:
        return Signal.SELL
 
    if (cur_rsi > cfg.rsi_overbought or close1 > upper_k) and kama_color == 2:
        return Signal.BUY
 
    return Signal.NONE
 
 
# ---------------------------------------------------------------------------
# Public signal function — SignalRSIExhaustion (orchestrator)
# ---------------------------------------------------------------------------
 
def signal_rsi_exhaustion(rates: pd.DataFrame) -> Signal:
    """
    Main entry point. Call once per new closed bar.
 
    Returns Signal.BUY / Signal.SELL / Signal.NONE.
    """
    data = _update_mr_indicators(rates)
    if data is None:
        return Signal.NONE
 
    if not _pass_re_atr_compression_filter(rates):
        return Signal.NONE
 
    return _detect_rsi_exhaustion(
        data["cur_rsi"],
        data["upper_k"],
        data["lower_k"],
        data["kama_color"],
        data["close1"],
    )

File: King of Modules/Zorro/Backups/19-03-2026 v1.0.0/Strategies/RSIExhaustion.c
// ============================================================
//  Strategies/RSIExhaustion.c
//  Zorro translation of Strategies/RSIExhaustion.mqh
//  Original: EA_KingofModules v1.30
//  Authors:  Daniel Pereira & Lucas Mattos
// ============================================================
//
//  ARCHITECTURE NOTES (MQL5 → Zorro)
//  ------------------------------------------------------------------
//  Bar-change guard (lastBar / iTime) → REMOVED (Zorro per-bar).
//
//  UpdateMRIndicators() shared with MeanReversion.mqh in MQL5.
//  Here the same indicator values are obtained directly through
//  the canonical Indicators.c helpers.
//
//  PassREATRCompressionFilter:
//    Allows entry when ATR(14) * multiplier < ATR(50)
//    (price in a compressed, range-bound regime).
//    Note: UseATRMRTrendFilter guard is ABSENT in the original —
//    this filter always runs when the function is called.
//    Preserved as-is.
//
//  DetectRSIExhaustion:
//    MOMENTUM CONTINUATION strategy (opposite of MeanReversion):
//      Oversold/below lower Keltner + KAMA bearish → SELL (continue)
//      Overbought/above upper Keltner + KAMA bullish → BUY (continue)
//    KAMA color (Indicators.c convention):
//      1 = falling/bearish  → SELL condition  (was ==1 in MQL5 ✓)
//      0 = rising/bullish   → BUY  condition  (was ==2 in MQL5)
//
//  [WIRING NOTE] SignalRSIExhaustion is NOT called in the original
//    EntriesEngine(). It is implemented here fully and wired to
//    UseREStrategy in Entries.c via the placeholder comment.
//    Uncomment that block in Entries.c to activate.
// ============================================================

#ifndef RSI_EXHAUSTION_C
#define RSI_EXHAUSTION_C

#include "../Indicators.c"


// ============================================================
//  PassREATRCompressionFilter
//  Returns 1 when ATR(14)*multiplier < ATR(50) (compressed regime).
//  ATR(50) series is at a dedicated call-site here.
// ============================================================
static int PassREATRCompressionFilter()
{
    var atrFast = series(ATR(InpATRPeriod))[0];
    var atrSlow = series(ATR(50))[0];

    return (atrFast * InpATRRETrendMultiplier < atrSlow);
}


// ============================================================
//  DetectRSIExhaustion
//  Pure momentum-continuation logic:
//    SELL: extreme RSI/price on the low side + KAMA still falling
//    BUY:  extreme RSI/price on the high side + KAMA still rising
//
//  OR logic between RSI and Keltner conditions is preserved
//  faithfully from the original.
// ============================================================
static ENUM_SIGNAL DetectRSIExhaustion(var curRSI,
                                        var upperK, var lowerK,
                                        int kamaColor, var close1)
{
    // SELL: oversold OR below lower band, KAMA bearish (falling)
    if((curRSI < InpRSIOversold || close1 < lowerK) && kamaColor == 1)
        return SIGNAL_SELL;

    // BUY:  overbought OR above upper band, KAMA bullish (rising)
    if((curRSI > InpRSIOverbought || close1 > upperK) && kamaColor == 0)
        return SIGNAL_BUY;

    return SIGNAL_NONE;
}


// ============================================================
//  SignalRSIExhaustion  (public)
// ============================================================
ENUM_SIGNAL SignalRSIExhaustion()
{
    // ── Gather indicators ────────────────────────────────────
    var curRSI = 0, prevRSI = 0;
    if(!GetRSI(&prevRSI, &curRSI))
        return SIGNAL_NONE;

    var upperK = 0, lowerK = 0;
    if(!GetKeltner(&upperK, &lowerK))
        return SIGNAL_NONE;

    int kamaColorPrev = 0, kamaColorCur = 0;
    if(!GetKAMAColor(&kamaColorPrev, &kamaColorCur))
        return SIGNAL_NONE;

    var close1 = seriesClose()[1];   // bar[-1] close

    // ── Filter ───────────────────────────────────────────────
    if(!PassREATRCompressionFilter())
        return SIGNAL_NONE;

    // ── Signal ───────────────────────────────────────────────
    return DetectRSIExhaustion(curRSI, upperK, lowerK,
                                kamaColorCur, close1);
}

#endif // RSI_EXHAUSTION_C
